### PR TITLE
feat: added cipher and derive methods for Symbol

### PR DESCRIPTION
**Pull request by rg911**

### CODE REVIEW

**Review by gimre-xymcity — DISMISSED** [review]

File: src/core/symbol/SymbolCrypto.ts [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+/**
+ * Prepare scalar mulResult
+ * @param privateKey - Private key.
+ * @returns - Prepared scalar mulResult
+ */
+const prepareForScalarMult = (privateKey: Uint8Array): Uint8Array => {
+    const d = new Uint8Array(64);
+    const hash = sha512.arrayBuffer(privateKey);
+    d.set(new Uint8Array(hash));
```
Comment (Jaguar0625) [inline src/core/symbol/SymbolCrypto.ts:32]: why are you calling set instead of just creating d here?

File: src/core/symbol/SymbolCrypto.ts [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+/***
+ * Encode a message, separated from encode() to help testing
+ *
+ * @param privateKey - Sender's private key.
+ * @param otherPublicKey - Recipient's public key.
+ * @param message - Message to be encoded.
+ * @param isMessageHex - Is message in hexadecimal format or not.
+ * @returns - Encoded message bytes
+ */
+export const encode = (privateKey: Key, otherPublicKey: Key, message: string, isMessageHex = false, customIv?: Uint8Array): Uint8Array => {
```
[the remark this answers is not in the document]
Comment (rg911) [reply src/core/symbol/SymbolCrypto.ts:49]: agree, should be uin8

File: src/core/symbol/external/nacl_catapult.ts [hunk 3]
```diff
@@ -0,0 +1,781 @@
+/*
```
[the remark this answers is not in the document]
Comment (rg911) [reply src/core/symbol/external/nacl_catapult.ts:1]: Question @Jaguar0625:
the tweetnacl package don't have the functions we used to derive shared secret exposed.  using the package's scalarmult(n,p) would produce different results?

[the remark this answers is not in the document]
Comment (Jaguar0625) [reply src/core/symbol/external/nacl_catapult.ts:1]: ok, we can skip this for now, but add an issue for it. afaik, the only reason we have special nacl file for NEM is because we're using a non-standard hash.

File: test/BasicVectorTest.template.ts [hunk 4]
```diff
@@ -94,3 +93,48 @@ export const AddressMosaicIdTester = <T extends Network>(
...
+                    Key.createFromHex(item.otherPublicKey).toBytes(),
+                );
+                const sharedKey = deriveSharedKey(
+                    Key.createFromHex(item.privateKey).toBytes(),
+                    Key.createFromHex(item.otherPublicKey).toBytes(),
+                );
+                // Assert:
+                const message = ` from ${item.privateKey}`;
+                expect(Converter.uint8ToHex(sharedSecret), `ScalarMulResult ${message}`).equal(item.scalarMulResult);
+                expect(Converter.uint8ToHex(sharedKey), `SharedKey ${message}`).equal(item.sharedKey);
```
Comment (Jaguar0625) [inline test/BasicVectorTest.template.ts:135]: mul = mult?

File: test/core/symbol/SymbolCrypto.spec.ts [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+import { expect } from 'chai';
+
+describe('Symbol Crypto', () => {
+    const sender = new SymbolKeyPair(Key.createFromHex('E1C8521608F4896CA26A0C2DE739310EA4B06861D126CF4D6922064678A1969B'));
+    const recipient = new SymbolKeyPair(Key.createFromHex('A22A4BBF126A2D7D7ECE823174DFD184C5DE0FDE4CB2075D30CFA409F7EF8908'));
+    const messageBytes = Converter.utf8ToUint8('NEM is awesome !');
+    const customerIV = Converter.hexToUint8('1D253DAF5C080DF0A3FD3F82');
+    const emptyKey = new Key(new Uint8Array());
+
+    const encodeAndDecodeAssert = (customerIV?: Uint8Array): Uint8Array => {
```
Comment (Jaguar0625) [inline test/core/symbol/SymbolCrypto.spec.ts:27]: customer => custom ?

File: test/core/symbol/SymbolCrypto.spec.ts [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+        it('Decode throws if using empty key', () => {
+            // Arrange:
+            const encoded = encodeAndDecodeAssert();
+            // Act+Assert:
+            expect(() => {
+                decode(recipient.privateKey, emptyKey, encoded);
+            }).to.throw();
+            expect(() => {
+                encode(sender.privateKey, emptyKey, messageBytes);
+            }).to.throw();
```
Comment (Jaguar0625) [inline test/core/symbol/SymbolCrypto.spec.ts:87]: decode?

File: test/core/symbol/SymbolCrypto.spec.ts [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+                encode(sender.privateKey, recipient.publicKey, messageBytes, new Uint8Array(1));
+            }).to.throw();
+        });
+
+        it('decode throws if using invalid payload', () => {
+            // Arrange:
+            const encoded = encodeAndDecodeAssert();
+
+            // Act+Assert:
+            expect(encoded.length).greaterThan(28);
```
Comment (Jaguar0625) [inline test/core/symbol/SymbolCrypto.spec.ts:109]: not sure purpose of this ?

File: test/core/symbol/SymbolCrypto.spec.ts [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+            expect(() => {
+                decode(recipient.privateKey, sender.publicKey, new Uint8Array(10));
+            }).to.throw();
+        });
+    });
+});
```
Comment (Jaguar0625) [inline test/core/symbol/SymbolCrypto.spec.ts:115]: would probably group decode and encode edge cases together

File: test/core/symbol/SymbolCrypto.spec.ts (hunk 7 above)
Comment (rg911) [reply test/core/symbol/SymbolCrypto.spec.ts:109]: To make sure the payload contains a 12 bytes IV and the 16 bytes tag?

File: test/BasicVectorTest.template.ts (hunk 4 above)
Comment (rg911) [reply test/BasicVectorTest.template.ts:135]: image

Yeah, the method name is using the `mult`, but the vector json key is using `mul` I was trying to make it consistent

**Review by Jaguar0625 — APPROVED** [review]
